### PR TITLE
feat(v2): add custom props for consumption by swizzled sidebar

### PR DESCRIPTION
**Pull request by oriooctopus**
## Motivation

-I want to pass in custom props for consumption by my swizzled DocSidebar component

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

I modified a test within `docusaurus-plugin-content/src/__tests__/sidebars.test.ts`

## Related PRs

https://github.com/facebook/docusaurus/issues/3868

### CODE REVIEW

File: packages/docusaurus-plugin-content-docs/src/plugin-content-docs.d.ts [hunk 1]
```diff
@@ -21,18 +21,22 @@ declare module '@docusaurus/plugin-content-docs-types' {
...
+  };
+
+  export interface PropSidebarItemLink extends PropsSidebarItemBase {
     type: 'link';
     href: string;
     label: string;
-  };
+  }
 
-  export type PropSidebarItemCategory = {
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/plugin-content-docs.d.ts:30]: Can you avoid this refactor please, and keep the existing type aliases instead of changing to interface?

You can do `export PropSidebarItemCategory = PropsSidebarItemBase  & {...}`

File: packages/docusaurus-plugin-content-docs/src/props.ts [hunk 2]
```diff
@@ -39,6 +39,7 @@ Available document ids=
       type: 'link',
       label: sidebar_label || title,
       href: permalink,
+      extra: item.extra,
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/props.ts:42]: Personally, I like "data" more for the name of this attribute, but no strong opinion

@ReginaLiang what's your opinion?

File: website/docs/docs.md [hunk 3]
```diff
@@ -165,6 +165,8 @@ As the name implies, `SidebarItem` is an item defined in a Sidebar. There are a
 - [Ref](#ref)
 - [Category](#category)
 
+**Note**: If you would like to pass in custom props to a swizzled sidebar, an optional object called `extra` can be added to any of the items:
```
Comment (slorber) [inline website/docs/docs.md:168]: That does not seem good enough to me, it looks like you forgot to add an example after the `:` ?

**Review by slorber — CHANGES_REQUESTED** [review]
The implementation looks fine, but please avoid refactoring types unnecessarily, that's outside the scope of this feature request.

File: packages/docusaurus-plugin-content-docs/src/props.ts (hunk 2 above)
Comment (ReginaLiang) [reply packages/docusaurus-plugin-content-docs/src/props.ts:42]: @slorber, 
Actually, I think any one will be fine. ^v^, a little little thought, How about `customFields`?  It is consistent with DocusaurusConfig object,  I guess `customFields` more meaningful. But any one is OK for me.

@slorber, @oriooctopus ,
Thanks for your guys great work! I believe docusaurus is getting better and better.

**Comment by oriooctopus** [comment]
@slorber would you like an additional commit with the changes or would you prefer them squashed/amended?

**Comment by slorber** [comment]
@oriooctopus it's not needed to squash because we squash all prs when merging, so we don't care about your local history 😉

**Comment by oriooctopus** [comment]
@slorber does this look good to you?

File: packages/docusaurus-plugin-content-docs/src/sidebars.ts [hunk 4]
```diff
@@ -20,7 +21,7 @@ import {
 import {mapValues, flatten, difference} from 'lodash';
 import {getElementsAround} from '@docusaurus/utils';
 
-type SidebarItemCategoryJSON = {
+export type SidebarItemCategoryJSON = SidebarItemBase & {
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/sidebars.ts:24]: does it need an export? I don't see it imported anywhere, why this change?

File: packages/docusaurus-plugin-content-docs/src/types.ts [hunk 5]
```diff
@@ -75,23 +75,27 @@ export type PluginOptions = MetadataOptions &
...
+};
+
+export interface SidebarItemDoc extends SidebarItemBase {
   type: 'doc' | 'ref';
   id: string;
-};
+}
 
-export type SidebarItemLink = {
+export interface SidebarItemLink extends SidebarItemBase {
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/types.ts:87]: same as before: use type aliases instead of refactoring to interface

File: packages/docusaurus-plugin-content-docs/src/plugin-content-docs.d.ts [hunk 6]
```diff
@@ -21,13 +21,17 @@ declare module '@docusaurus/plugin-content-docs-types' {
     permalinkToSidebar: PermalinkToSidebar;
   };
 
-  export type PropSidebarItemLink = {
+  export type PropsSidebarItemBase = {
+    extra?: object;
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/plugin-content-docs.d.ts:25]: now renamed to customProps?

**Review by slorber — CHANGES_REQUESTED** [review]
Still some type->interface refactorings
+ "extra" still there after the renaming.

Otherwise LGTM. 
Not sure `customProps` is the best name, but let's merge this. We'll see later if anyone can come with a better name 😅

**Comment by oriooctopus** [comment]
Just updated it, sorry about the changes I missed last time. 
Question: what's the reason for not importing types in plugin-content-docs.d.ts but importing them in sidebars.ts? Excuse me if this is a simple question, I'm new to typescript

**Comment by slorber** [comment]
thanks, LGTM 👍 

@oriooctopus the TS setup is a bit messy currently, not sure to have a good answer for that, but we'll try to improve it over time :)
Initially the codebase was mostly JS so it's an incremental upgrade
